cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/ProgramScheduleStoreDataset.java

```diff
@@ -375,6 +378,57 @@ public void deleteSchedules(Iterable<? extends ScheduleId> scheduleIds) throws N
   }
 
   /**
+   * Update all schedules that can be triggered by the given deleted program. Schedules will be removed if they
+   * contain single {@link co.cask.cdap.internal.app.runtime.schedule.trigger.ProgramStatusTrigger}. Schedules with
+   * composite triggers will be updated if the composite trigger can still be satisfied after the program is deleted,
+   * otherwise the schedules will be deleted.
+   *
+   * @param programId the program id for which to delete the schedules
+   * @return the IDs of the schedules that were deleted
+   */
+  public List<ScheduleId> modifySchedulesTriggeredByDeletedProgram(ProgramId programId) {
+    List<ScheduleId> deleted = new ArrayList<>();
+    Set<ProgramScheduleRecord> scheduleRecords = new HashSet<>();
+    for (ProgramStatus status : ProgramStatus.values()) {
+      scheduleRecords.addAll(findSchedules(Schedulers.triggerKeyForProgramStatus(programId, status)));
+    }
+    for (ProgramScheduleRecord scheduleRecord : scheduleRecords) {
+      ProgramSchedule schedule = scheduleRecord.getSchedule();
+      try {
+        deleteSchedule(schedule.getScheduleId());
+      } catch (NotFoundException e) {
+        // this should never happen
+        LOG.warn("Failed to delete the schedule '{}' triggered by '{}', skip this schedule.",
+                 schedule.getScheduleId(), programId, e);
+        continue;
+      }
+      if (schedule.getTrigger() instanceof AbstractCompositeTrigger) {
+        // get the updated composite trigger by removing the program status trigger of the given program
+        Trigger updatedTrigger = ((AbstractCompositeTrigger) schedule.getTrigger())
+          .updateTriggerWithDeletedProgram(programId);
+        if (updatedTrigger == null) {
+          deleted.add(schedule.getScheduleId());
+          continue;
+        }
+        // if the updated composite trigger is not null, add the schedule back with updated composite trigger
+        try {
+          addScheduleWithStatus(new ProgramSchedule(schedule.getName(), schedule.getDescription(),
+                                                    schedule.getProgramId(), schedule.getProperties(), updatedTrigger,
+                                                    schedule.getConstraints(), schedule.getTimeoutMillis()),
+                                scheduleRecord.getMeta().getStatus(), System.currentTimeMillis());
+        } catch (AlreadyExistsException e) {
+          // this should never happen
+          LOG.warn("Failed to add the schedule '{}' triggered by '{}' with updated trigger '{}', " +
```

Comment:
Seems like this can happen if program C is triggered by program A or program B and program A and program B are deleted at the same time. In that scenario, you'll end up with an incorrect schedule right? Unless this entire method is running in a transaction?